### DISCUSSION TITLE
Missing libzkgroup on v0.7.1

### DISCUSSION THREAD

**Discussion by DutchForeigner** [AsamK/signal-cli#393] [Q&A] [answered]
Hi,
I'm getting the below error when trying to send an message:
"OpenJDK Server VM warning: You have loaded library /tmp/resource14973433511044593591.so which might have disabled stack guard. The VM will try to fix the stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
WARNING: Support for new group V2 is disabled, because the required native library dependency is missing: libzkgroup"

Initially I tried to use  https://github.com/AsamK/signal-cli/releases/download/v0.7.1/signal-cli-0.7.1.tar.gz
but as that was giving above error, I also tried to build it from source with Gradle v4.4.1, and that gave also the same error.

Looking at bug report See also "[Bug 252041](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=252041) - net-im/signal-cli: Update to 0.7.1" I thought it would have been fixed in a newer version but no luck so far....

Using:
Raspberry Pi 3, Raspbian Linux 10 (buster) 
Java JDK 11.0.9.1 
Gradle v4.4.1

**Comment by AsamK** [comment]
See https://github.com/AsamK/signal-cli/wiki/Provide-native-lib-for-libsignal

**Comment by DutchForeigner** [comment, marked as answer]
Thanks for that,
Hope you had a merry Christmas and wishing you a Happy New year/ Happy holiday season.

In the end it was not _that_ straight forward…

So as FYI: below are the steps which worked for me, on a Raspberry Pi 3B with Raspbian GNU/Linux 10 (buster)
My main user is 'smarthome', replace it with 'pi' if that is still the main user.

### Install required packages to be able to build signal-cli and dependent zklibrary
[smarthome@1234 ~]$ **mkdir -p /home/smarthome/_downloaded**

### Install latest JAVA JDK (v11.0.9.1)
[smarthome@1234 ~]$ **sudo apt update**
[smarthome@1234 ~]$ **sudo apt install default-jdk**
[smarthome@1234 ~]$ **java --version**

### Install latest version (v6.7.1) of Gradle 
(https://gradle.org/install/) 
Copy down binary-only from: https://gradle.org/next-steps/?version=6.7.1&format=bin
[smarthome@1234 ~]$ **wget -P /home/smarthome/_downloaded https://services.gradle.org/distributions/gradle-6.7.1-bin.zip**

[smarthome@1234 ~]$ **sudo mkdir /opt/gradle**
[smarthome@1234 ~]$ **sudo unzip -d /opt/gradle /home/smarthome/_downloaded/gradle-6.7.1-bin.zip**

Add gradle to the PATH:
[smarthome@1234 ~]$ **export PATH=$PATH:/opt/gradle/gradle-6.7.1/bin**

Check if all is ok by getting the version of Gradle:
[smarthome@1234 ~]$ **gradle -v**
------------------------------------------------------------
Gradle 6.7.1
------------------------------------------------------------

### Install Rust:
https://www.rust-lang.org/tools/install
Run the below and follow instructions:
[smarthome@1234 ~]$ **curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh**
 **Choose 1**  Proceed with installation (default)

Once finished, add rust to the Path
[smarthome@1234 ~]$ **export PATH="$PATH:/home/smarthome/.cargo/bin"**

Check rust is working by getting its version:
[smarthome@Home ~]$ **rustup --version**
rustup 1.23.1 (3df2264a9 2020-11-30)
info: This is the version for the rustup toolchain manager, not the rustc compiler.

### Get and build zkgroup library:
Download the Source code for zkgroup library v0.7.1 (https://github.com/signalapp/zkgroup/releases)
[smarthome@1234 ~]$ **wget -O /home/smarthome/_downloaded/zkgroup-0.7.1.zip https://github.com/signalapp/zkgroup/archive/v0.7.1.zip**

Extract content of zkgroup-0.7.1.zip to /home/smarthome:
[smarthome@1234 ~]$ **unzip -d /home/smarthome /home/smarthome/_downloaded/zkgroup-0.7.1.zip**

[smarthome@1234 ~]$ **cd /home/smarthome/zkgroup-0.7.1**
Check rustc version:
[smarthome@1234 ~]$ **rustc --version**
rustc 1.41.1 (f3e1a954d 2020-02-24)
[smarthome@1234 ~]$ **make libzkgroup**     (note this takes several minutes (~6) to complete)
Once finished, the required file is:  /home/smarthome/zkgroup-0.7.1/target/release/libzkgroup.so

### Build signal-cli
https://github.com/AsamK/signal-cli "Building" section:
[smarthome@1234 ~]$ **mkdir /home/smarthome/signal-cli_source**
[smarthome@1234 ~]$ **cd /home/smarthome/signal-cli_source**
[smarthome@1234 ~]$ **git clone https://github.com/AsamK/signal-cli.git**
[smarthome@1234 ~]$ **cd ./signal-cli**
[smarthome@1234 ~]$ **gradle build**   (note this takes several minutes (~10) to complete)
[smarthome@1234 ~]$ **gradle installDist**
[smarthome@1234 ~]$ **gradle distTar**

### Update jar file with updated libzkgroup.so file
[smarthome@1234 ~]$ **cd  /home/smarthome/zkgroup-0.7.1/target/release/**
[smarthome@1234 ~]$ **zip -u /home/smarthome/signal-cli_source/signal-cli/build/install/signal-cli/lib/zkgroup-java-0.7.0.jar libzkgroup.so**

### Copy JAR File
[smarthome@1234 ~]$ **cd /home/smarthome/signal-cli_source/signal-cli/build/distributions**
[smarthome@1234 ~]$ **mkdir -p signal-cli-0.7.1/lib/**
[smarthome@1234 ~]$ c**p /home/smarthome/signal-cli_source/signal-cli/build/install/signal-cli/lib/zkgroup-java-0.7.0.jar signal-cli-0.7.1/lib/**

### Update ZIP file
[smarthome@1234 ~]$ **zip -u  /home/smarthome/signal-cli_source/signal-cli/build/distributions/signal-cli-0.7.1.zip signal-cli-0.7.1/lib/zkgroup-java-0.7.0.jar**

### Update TAR file
[smarthome@1234 ~]$ **tar --delete -vPf /home/smarthome/signal-cli_source/signal-cli/build/distributions/signal-cli-0.7.1.tar signal-cli-0.7.1/lib/zkgroup-java-0.7.0.jar**
[smarthome@1234 ~]$ **tar --owner='' --group='' -rvPf /home/smarthome/signal-cli_source/signal-cli/build/distributions/signal-cli-0.7.1.tar signal-cli-0.7.1/lib/zkgroup-java-0.7.0.jar**

Updated signal-cli-0.7.1.tar is now in:  /home/smarthome/signal-cli_source/signal-cli/build/distributions/

### Install/export signal-cli
[smarthome@1234 ~]$ **sudo tar xf /home/smarthome/signal-cli_source/signal-cli/build/distributions/signal-cli-0.7.1.tar -C /opt**
[smarthome@1234 ~]$ **sudo ln -sf /opt/signal-cli-0.7.1/bin/signal-cli /usr/local/bin/**

- signal-cli is now installed on the Raspberry Pi
Issue can be closed.
:-)

**Reply by jliebers** [reply]
Thank you so much! I struggled with the same problem on my raspberry and your post helped me to a huge extent. Thanks! :)

**Reply by Atomic-Bean** [reply]
I'll add a notice to the original page that you'll need to manually build it yourself for PIs and other Archs

**Reply by hennjo** [reply]
Hi there,

thank you for your detailed instructions. I tried to follow them on a rpi3b+ with 1 GB of RAM. This pi wasn't able to build the signal-cli instance because it ran out of RAM.
After that i used a rpi4 with 8 GB of RAM and it got it done, but gives me the following error:
`signal-cli -u +ZZYYYXXXXXXXX register --voice
OpenJDK Server VM warning: You have loaded library /tmp/resource18301514204854240517.so which might have disabled stack guard. The VM will try to fix the stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
ERROR App - Error loading or creating state file: /tmp/resource18301514204854240517.so: /tmp/resource18301514204854240517.so: wrong ELF class: ELFCLASS64 (Possible cause: architecture word width mismatch)`
The only difference is, that I used the 0.7.4 version of signal-cli

Does anybody have some insight into a solution?

In the meantime I'll try the file you uploaded below. Saw that only after logging into github.

**Reply by bruno00o** [reply]
hi,

at first, i had the same issue as you on my rpi3. i don't have an rpi4 so i've done @DutchForeigner's instructions several times and had the same problem as you on your rpi4.

the only way i managed to properly install signal-cli is by using @DutchForeigner's TAR file on his [github](https://github.com/DutchForeigner/signal-cli_rpi) 
you'll get signal-cli v0.7.1 but it works!

**Reply by davemacrae** [reply]
After trying to get @DutchForeigner's code to compile on my Rpi4 I had similar issues with the ELF error.

However, when I just replaced the distributed zkgroup file with one I build using the instructions above, everything worked well, i.e. I have signal-cli V0.7.4 running and being able to access V2 groups.

I've placed the file in my Git [repo](https://github.com/davemacrae/signal-zkgroup) should anyone want to use it.

**Reply by benni2108** [reply]
I have the same problem. But sadly your solution doesn't work for me. The Problem is that signal-cli 0.7.1 is not functioning and with 0.8.1 i get a "Missing required native library dependency: libsignal-client" error.

Do you have any ideas what could possibly be the problem. How do I add the depenency?

**Reply by Atomic-Bean** [reply]
The "libsignal-client" error is referring to this sort of stuff. https://github.com/AsamK/signal-cli/wiki/Provide-native-lib-for-libsignal

**Comment by bruno00o** [comment]
thank you for your work @DutchForeigner! i tried to do same on rpi 3 too but it crashes everytime i do the gradle buil... Does anyone know an alternative or a solution pls?

**Reply by DutchForeigner** [reply]
I've now posted the TAR file, I build using the earlier mentioned steps on my Raspberry PI 3, on Github.
Hope that one will work for you.
[Github link](https://github.com/DutchForeigner/signal-cli_rpi)

**Reply by bruno00o** [reply]
it worked!! thank you so much! maybe it was because of the newest version of signal-cli

**Reply by DutchForeigner** [reply]
Hi, signal-cli v0.7.4 tar file, for on the Raspberry Pi, is now also on my Github page.
[Gitub link](https://github.com/DutchForeigner/signal-cli_rpi)

**Reply by DutchForeigner** [reply]
Hi, signal-cli v0.8.1 tar file, for on the Raspberry Pi, is now also on my Github page, with updated compile 'manual'
[Gitub link](https://github.com/DutchForeigner/signal-cli_rpi)
